Improve handling of null SAMFileHeaders in alignment records.

cmnbroad opened the pull request:
- This is the first pass of improving null SAMFileHeader support. There are some TODOs inline with questions.
- Once we have agreement on those, the next pass will be to find all the places that call getReferenceIndex/getReferenceName and the mate analogs internally and make sure they do something reasonable with the new sentinel values.
- The existing doc seems to already indicate that the header does not have to be set. We may want to add more detail there, but I've updated the doc for the individual methods, and will clean those up
again once we have resolution on the rest of the contract (ie null value handling, etc.)
- I chose to return a MISSING_HEADER validation error when validating records. Not sure if thats what  we want.
- The existing implementation silently allows the caller to set an invalid reference name (one that doesn't exist in the sequence dictionary), but throws if you try to set an invalid reference index. I've retained that behavior but am wondering if its deliberate/correct.
- I don't see a code path that requires any changes to BAMRecord. Likeiwse there are numerous code paths that assume non null headers that I didn't touch because the records all appear to originate from files and have valid headers (i.e., SAMFileValidator).
- The existing doc says get/setReferenceIndex() is preferred to get/setReferenceName(). Should we change that ?

droazen commented:
We discussed this PR in person and agreed on a set of changes to be made before an in-depth review.

cmnbroad commented:
Summary of behavior with these changes:

SAMRecord:
-setHeader(null) clears the reference indices and sets them to null (leaving them unresolved). Setting the header to non-null attempts to re-resolve the reference names against the current dictionary and updates the indices. This can fail silently in the same way that setReferenceName can fail silently in the existing implementation.
-getReferenceName() is always safe (as before).
-setReferenceName(name) is always safe if the name is non-null (as before). If the record has a header, it will attempt to resolve the name to an index, but this may silently fail and leave the index unresolved.
-setReferenceIndex(NO_ALIGNMENT_INDEX) is always safe. Using any other value will throw if the record has no header or if the value can't be found in the dictionary.
-getReferenceIndex() is unsafe if the record has no header, otherwise it is safe, though if the index is
unresolved and the reference name can't be resolved to an index using the current dictionary, the index will be clobbered and reset to NO_ALIGNMENT_INDEX (as before!!)

Other:
-The SAM/BAM/CRAM writers all now have the side effect that they will  reset a record's header when handed a record that doesn't already have one.
-Note: With these changes, it is not possible to set a SAMRecord's header to null and then directly (without going through a writer) encode it to a stream using the BAMRecordCodec, since setting the
header to null clobbers the reference indices. (This will require a change to the Hellbender read serializer test since it tries to do this).

droazen commented on src/java/htsjdk/samtools/DefaultSAMRecordFactory.java:
```diff
@@ -18,7 +18,11 @@ public SAMRecord createSAMRecord(final SAMFileHeader header) {
-    /** Create a new BAM Record. */
+    /**
+     * Create a new BAM Record. If the reference sequence index or mate reference sequence index are
+     * any value other than NO_ALIGNMENT_REFERENCE, the values must be resolvable against the sequence
```
`NO_ALIGNMENT_REFERENCE_INDEX`

droazen commented on src/java/htsjdk/samtools/MergingSamRecordIterator.java:
```diff
@@ -144,17 +145,6 @@ public SAMRecord next() {
-            if (record.getReadPairedFlag() && record.getMateReferenceIndex() != SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX) {
-                record.setMateReferenceIndex(this.samHeaderMerger.getMergedSequenceIndex(iterator.getReader().getFileHeader(), record.getMateReferenceIndex()));
-            }
-        }
```
Are we 100% sure that we can delete this block? I know that the header merger uses per-header translation tables mapping from old to new indices -- does the `record.setHeader(this.samHeaderMerger.getMergedHeader());` line above render this unnecessary?

cmnbroad replied:
I don' think we can leave this code in with the current changes - it assumes the reference indices are stale and relative to the old dictionary, and tries to resolve them by translation. With these changes, the setHeader call above will resolve the indices (by name) against the new dictionary, so they'll already be translated. Treating them as if they're in the old space would be incorrect I think. I should be able to add a test to verify this.

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
 
     /**
-     * @return Reference name, or null if record has no reference.
+     * @return Reference name, or NO_ALIGNMENT_REFERENCE_NAME (-1) if the record has no reference name
```
Value of `NO_ALIGNMENT_REFERENCE_NAME` is `"*"`, not `-1`

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
                 mReferenceIndex = mHeader.getSequenceIndex(mReferenceName);
             }
+            else {
+                throw new IllegalArgumentException("A non-null SAMFileHeader is required to resolve the reference index");
```
This should be an `IllegalStateException`, I think (there are no arguments to this method).

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
     /**
-     * @param referenceIndex Must either equal -1 (indicating no reference), or exist in the sequence dictionary
-     * in the header associated with this record.
+     * Updates the reference index. The record must have a valid SAMFileHeader, and the reference index must appear in
```
"The record must have a valid SAMFileHeader, unless the referenceIndex parameter equals NO_ALIGNMENT_REFERENCE_INDEX"

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
-                throw new IllegalArgumentException("Reference index " + referenceIndex + " not found in sequence dictionary.", e);
+        }
+        else if (null == mHeader) {
+            throw new IllegalArgumentException("A non-null SAMFileHeader is required to resolve the reference index");
```
`IllegalStateException`

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
+    /**
+     * Sets the reference name for this record. If the record has a valid SAMFileHeader and the reference
+     * name is present in the associated sequence dictionary, the record's reference index will also be
+     * updated with the corresponding sequence index.
```
"If `referenceName` is `NO_ALIGNMENT_REFERENCE_NAME`, sets the reference index to `NO_ALIGNMENT_REFERENCE_INDEX`

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
+     *
+     * If the reference name for this record has previously been resolved against the sequence dictionary, the corresponding
+     * index is returned directly. Otherwise, the record must have a non-null SAMFileHeader that can be used to
+     * resolve the index for the record's current reference name. If the record has a header, and the name does not appear
```
unless the reference name is `NO_ALIGNMENT_REFERENCE_NAME`

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
 
     /**
-     * @return Mate reference name, or null if one is not assigned.
+     * @return Reference name, or NO_ALIGNMENT_REFERENCE_NAME (-1) if the record has no reference name
```
`NO_ALIGNMENT_REFERENCE_NAME ("*")`

droazen replied:
"reference name" -> "mate reference name" (at both the beginning and end of this line)

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
+    /**
+     * Sets the mate reference name for this record. If the record has a valid SAMFileHeader and the mate reference
+     * name is present in the associated sequence dictionary, the record's mate reference index will also be
+     * updated with the corresponding sequence index.
```
"If mateReferenceName is NO_ALIGNMENT_REFERENCE_NAME, sets the mate reference index to NO_ALIGNMENT_REFERENCE_INDEX"

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
+     *
+     * If the mate reference name for this record has previously been resolved against the sequence dictionary, the
+     * corresponding index is returned directly. Otherwise, the record must have a non-null SAMFileHeader that can be
+     * used to resolve the index for the record's current mate reference name. If the record has a header, and the
```
unless the mate reference name is NO_ALIGNMENT_REFERENCE_NAME

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
                 mMateReferenceIndex = mHeader.getSequenceIndex(mMateReferenceName);
             }
+            else {
+                throw new IllegalArgumentException("A non-null SAMFileHeader is required to resolve the mate reference index");
```
`IllegalStateException`

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
-    public void setMateReferenceIndex(final int referenceIndex) {
-        mMateReferenceIndex = referenceIndex;
-        if (mMateReferenceIndex == NO_ALIGNMENT_REFERENCE_INDEX) {
+     * Updates the mate reference index. The record must have a valid SAMFileHeader, and the mate reference index must appear in
```
"The record must have a valid SAMFileHeader, unless the mateReferenceIndex parameter equals NO_ALIGNMENT_REFERENCE_INDEX"

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -305,119 +308,202 @@ private boolean hasMateReferenceName() {
-                throw new IllegalArgumentException("Reference index " + referenceIndex + " not found in sequence dictionary.", e);
+        }
+        else if (null == mHeader) {
+            throw new IllegalArgumentException("A non-null SAMFileHeader is required to resolve the mate reference index");
```
`IllegalStateException`

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -684,15 +770,15 @@ protected void initializeCigar(final Cigar cigar) {
      * @return The SAMReadGroupRecord from the SAMFileHeader for this SAMRecord, or null if
      * 1) this record has no RG tag, or 2) the header doesn't contain the read group with
      * the given ID.
```
"or 3) this record has no SAMFileHeader"

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1337,16 +1423,70 @@ int computeIndexingBin() {
+    /**
+     *
+     * @return the SAMFileHeader for this record. If the header is null, the following SAMRecord methods may throw
+     * IllegalArgumentExceptions:
```
`IllegalArgumentExceptions" -> "exceptions" (since I asked you to change several exception classes above)

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1337,16 +1423,70 @@ int computeIndexingBin() {
+     *   setMateReferenceIndex
+     *
+     * A SAMRecord with a null header can not be fully validated and will result in SAMValidationErrors if isValid is
+     * called.
```
Can you copy these warnings on `isValid()` and methods that can throw exceptions into a new paragraph on "Notes on headerless SAMRecords" in the class-level comment at the top of `SAMRecord`?

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1337,16 +1423,70 @@ int computeIndexingBin() {
+     * resolved and updated using the sequence dictionary in the new header.
+     *
+     * When the SAMFileHeader is set to null, the reference and mate reference indices are cleared. Therefore, calls to
+     * the following SAMRecord methods on records with a null header may throw IllegalArgumentExceptions:
```
"IllegalArgumentExceptions" -> "exceptions"

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1337,16 +1423,70 @@ int computeIndexingBin() {
+            }
+            if (null != mMateReferenceIndex && !mMateReferenceIndex.equals(NO_ALIGNMENT_REFERENCE_INDEX)) {
+                mMateReferenceIndex = null;
+            }
```
Is there a reason you're not clearing the indices if they're equal to `NO_ALIGNMENT_REFERENCE_INDEX`? I think you should be unconditionally nulling out the indices here for clarity's sake. If you null them out when they are equal to  `NO_ALIGNMENT_REFERENCE_INDEX`, and the corresponding reference name is truly equal to `NO_ALIGNMENT_REFERENCE_NAME`, the index will get reset to `NO_ALIGNMENT_REFERENCE_INDEX` anyway on the next access (even in the headerless case).

cmnbroad replied:
We're using the null value to indicate "unresolved", which is not quite correct in that case, but it seems harmless to do and cleans up the code so I'll change it.

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1337,16 +1423,70 @@ int computeIndexingBin() {
+        }
+        else {
+            // attempt to resolve the existing reference names and indices against the new sequence dictionary, but
+            // don't throw if the names don't appear in the dictionary
```
Make this a prominent warning in the method-level docs (that we won't throw if the existing reference names aren't present in the new header)

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1665,7 +1807,12 @@ public int hashCode() {
+                if (ret == null) ret = new ArrayList<SAMValidationError>();
+                ret.add(new SAMValidationError(SAMValidationError.Type.MISSING_HEADER, "A non-null SAMHeader is required to resolve the sequence dictionary for a mapped read.", getReadName()));
+                if (firstOnly) return ret;
+            }
```
Not sure I like this...the whole point of this PR was to make headerless `SAMRecords` an officially supported, valid state for a read to be in. Yet with this change they will fail `isValid()`. Can't we instead say that headerless `SAMRecords` are valid, but are validated to a less strict standard (ie., some checks that require a header are not performed)?

cmnbroad replied:
We could go either way (before this change, there was already a validation error enum value for MISSING_HEADER, thoughits unclear to me if there was actually any code path that would produce such an error on a record with a null header without first throwing an NPE). One other option I like is to change the MISSING_HEADER enum value to be a SAMValidationError.Severity.WARNING instead of an ERROR, and only return those for each case where a header is required for validation (including SAMRecord.validateCigar and SAMUtils.validateCigar which can both be called independently of SAMRecord.isValid()).

droazen commented on src/java/htsjdk/samtools/SAMRecordCoordinateComparator.java:
```diff
@@ -75,10 +75,16 @@ private int compareInts(int i1, int i2) {
      * Less stringent compare method than the regular compare.  If the two records
      * are equal enough that their ordering in a sorted SAM file would be arbitrary,
      * this method returns 0.  If read is paired and unmapped, use the mate mapping to sort.
+     * Record being compared must have non-null SAMFileHedaers.
```
Record -> Records

droazen replied:
SAMFileHedaers -> SAMFileHeaders

droazen commented on src/java/htsjdk/samtools/SAMRecordCoordinateComparator.java:
```diff
@@ -75,10 +75,16 @@ private int compareInts(int i1, int i2) {
     public int fileOrderCompare(final SAMRecord samRecord1, final SAMRecord samRecord2) {
+
+        if (null == samRecord1.getHeader() || null == samRecord2.getHeader()) {
+            throw new IllegalArgumentException("Record must have non-null SAMFileHeaders to be comapred");
```
Record -> Records
comapred -> compared

droazen commented on src/java/htsjdk/samtools/SAMRecordDuplicateComparator.java:
```diff
@@ -200,6 +204,10 @@ private boolean pairedEndAndBothMapped(final SAMRecord record) {
         int cmp;
 
+        if (null == samRecord1.getHeader() || null == samRecord2.getHeader()) {
+            throw new IllegalArgumentException("Record must have non-null SAMFileHeaders to be comapred");
```
Record -> Records
comapred -> compared

droazen replied:
Also this check is redundant, since the first thing `compare()` does is to call `fileOrderCompare()`

cmnbroad replied:
True that. Done.

droazen commented on src/java/htsjdk/samtools/SAMRecordDuplicateComparator.java:
```diff
@@ -236,6 +244,10 @@ public int compare(final SAMRecord samRecord1, final SAMRecord samRecord2) {
         int cmp;
 
+        if (null == samRecord1.getHeader() || null == samRecord2.getHeader()) {
+            throw new IllegalArgumentException("Record must have non-null SAMFileHeaders to be comapred");
```
Record -> Records
comapred -> compared

droazen commented on src/java/htsjdk/samtools/SAMUtils.java:
```diff
@@ -894,14 +900,22 @@ public static int getMateUnclippedEnd(final SAMRecord rec) {
+                if (ret == null) ret = new ArrayList<SAMValidationError>();
+                ret.add(new SAMValidationError(SAMValidationError.Type.MISSING_HEADER,
+                        cigarTypeName + " A non-null SAMHeader is required to validate cigar elements for: ", rec.getReadName(), recordNumber));
+            }
```
Here too, I think a missing header should result in skipped validation steps rather than a validation failure.

droazen commented on src/java/htsjdk/samtools/cram/build/Sam2CramRecordFactory.java:
```diff
@@ -96,7 +99,17 @@ public Sam2CramRecordFactory(final byte[] refBases, final SAMFileHeader samFileH
+     * Create a CramCompressionRecord.
+     *
+     * @param record If the input record does not have an associated SAMFileHeader, it will be updated
+     *               with the header used for the writer in order to allow reference indices to be resolved.
```
This class is not a "writer" class, is it?

cmnbroad replied:
Technically, no, though it only gets used to create records for CRAMFileWriter. Still, its confusing to use the word "writer" here so I'll change it to "factory."

droazen commented on src/java/htsjdk/samtools/util/AbstractProgressLogger.java:
```diff
@@ -76,7 +76,7 @@ public synchronized boolean record(final String chrom, final int pos) {
     public synchronized boolean record(final SAMRecord rec) {
-        if (rec.getReferenceIndex() == SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX) {
+        if (null == rec.getHeader() || rec.getReferenceIndex() == SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX) {
             return record(null, 0);
```
Recommend changing this method to use the reference name exclusively. Eg.,

    if ( SAMRecord.NO_ALIGNMENT_REFERENCE_NAME.equals(rec.getReferenceName()) ) {
        return record(null, 0);
    }
    else {
        return record(rec.getReferenceName(), rec.getAlignmentStart());
    }

droazen commented on src/tests/java/htsjdk/samtools/BAMFileWriterTest.java:
```diff
@@ -114,10 +118,72 @@ public void testPositive(final String testName, final SAMRecordSetBuilder samRec
+            final SAMRecord bamRecord = bamIt.next();
+
+            // test only reference names since we'll have lost reference indices due to the fake  null header
+            Assert.assertEquals(bamRecord.getReferenceName(), samRecord.getReferenceName());
```
Check equality of alignment starts here too

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+    @Test
+    public void testReferenceName() {
+        // NO_ALIGNMENT_NAME
+        SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
```
Extract this `new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);` call into a helper method, and call it here and everywhere else this line is duplicated.

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+    public void testInvalidReferenceIndex() {
+        // unresolvable reference
+        final SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        sam.setReferenceIndex(9999);
```
It's interesting that we throw on an invalid reference index, but not an invalid reference name (when there is a header). Perhaps for consistency's sake we should throw on `setReferenceName()` when a header is present and the reference name is not in the dictionary? Thoughts?

cmnbroad replied:
I agree its not ideal and if I was designing the API from scratch I wouldn't make that choice, but I wouldn't change it at this point either.

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+
+        sam.setHeader(null);
+        // setting header to null retains the previously assigned ref name
+        Assert.assertTrue(originalRefName.equals(sam.getReferenceName()));
```
Since `SAMRecord.mReferenceIndex` is `protected`, and you are in the same package, you should also be able to assert that the `setHeader(null)` call cleared the reference index.

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+    }
+
+    @Test
+    public void testNullHeaderReferenceIndex() {
```
`testNullHeaderReferenceName` and `testNullHeaderReferenceIndex` are not well decomposed -- they both appear to be testing reference name and index-related functionality in equal amounts. Recommend either merging them or better separating the testing of the reference name and index-related methods.

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+        sam.setHeader(null);
+        // setReferenceIndex with null header throws
+        sam.setReferenceIndex(3);
+    }
```
Add a companion test case that calls `getReferenceIndex()` on a `SAMRecord` with a null header, and expects an exception.

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+        // force the internal SAMRecord reference index value to null initial state
+        final SAMRecord sam = new SAMRecord(null);
+        sam.setReferenceName("unresolvable");
+        sam.getReferenceIndex();
```
Add a companion test case that starts with a fully-populated `SAMRecord` with valid contig, nulls out the header, calls `getReferenceIndex()`, and expects an exception.

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+        final String origMateRefName = sam.getMateReferenceName();
+        final int originalMateRefIndex = sam.getMateReferenceIndex();
+
+        // setting header to null retains ref name and index
```
It shouldn't retain the index, it should null it out (and you should check that it does, since `mMateReferenceIndex` has protected access and you're in the right package).

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+        // force the internal SAMRecord reference index value to null initial state
+        sam = new SAMRecord(null);
+        final String unresolvableRefName = "unresolvable";
+        sam.setMateReferenceName(unresolvableRefName);
```
Check value of `mMateReferenceIndex` at this point, taking advantage of protected access.

droazen commented on src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java:
```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+        final SAMRecord sam = new SAMRecord(null);
+        sam.setMateReferenceName("unresolvable");
+        sam.getMateReferenceIndex();
+    }
```
Add a test that starts with a sam with a header and a valid mate reference name, nulls the header, calls `getMateReferenceIndex()`, and expects an exception.

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -57,27 +57,30 @@
+ *
  * By default, extensive validation of SAMRecords is done when they are read.  Very limited validation is done when
  * values are set onto SAMRecords.
+ *
```
Add javadoc to the `SAMRecord` constructor documenting behavior when a null header is passed in (can re-use docs from `setHeader()`

droazen commented:
Review complete -- back to @cmnbroad. Looks very good overall, much closer to what we want than the first version. Mostly minor comments, but a few concerns here and there that need a response.

cmnbroad commented:
Back to @droazen. I made all of the changes except for how we handle validation errors; one other option is to make the null-header case a warning rather than a hard error (see comment above). Just wanted to present another option, let me know which you prefer and I'll make the code and doc changes for that.

I also added merging iterator test that forces and tests index remapping, and fixed up some javadoc. I'll also have to make a minor modification in this branch to the deepCopy code to accommodate these changes, but I can't do that until I rebase.

One more thing, I think I mentioned this a while back, but these changes break this [SAMRecordToGATKReadAdapterSerializer test ](https://github.com/broadinstitute/gatk/blob/master/src/test/java/org/broadinstitute/hellbender/engine/spark/SAMRecordToGATKReadAdapterSerializerUnitTest.java#L37) in Hellbender, since it tries to use the BAMRecordCodec to encode/decode records after their header has been set to null. The BAMRecordCodec doesn't serialize the ref names, only the indices, which are clobbered by the setHeader(null). The other cloud mandatory tests all pass.

cmnbroad commented:
Squashed and rebased in order to integrate deepCopy changes with null headers and test. Added a test to make sure SAMRecordMergingIterator handles index remapping properly. Validation with null headers is no longer a validation error. Ready for review once tests pass.

droazen commented:
Looks good, except that it appears that a headerless record still can't fully pass validation -- specifically, cigar validation via `SAMUtils.validateCigar()` will fail. This is a problem for the headerless use case, as `SAMRecord.getCigar()` calls `validateCigar()`, so we can't even get the cigar on a headerless record without getting an exception if STRICT validation is on. Can you address?
